programs/drift_vaults/src/state/vault.rs

```diff
@@ -500,4 +842,321 @@ impl Vault {
         self.liquidation_delegate = Pubkey::default();
         self.liquidation_start_ts = 0;
     }
+
+    pub fn protocol_request_withdraw(
+        &mut self,
+        vault_protocol: &mut Option<RefMut<VaultProtocol>>,
+        withdraw_amount: u64,
+        withdraw_unit: WithdrawUnit,
+        vault_equity: u64,
+        now: i64,
+    ) -> Result<()> {
+        self.apply_rebase(vault_protocol, vault_equity)?;
+        let VaultFee {
+            management_fee_payment,
+            management_fee_shares,
+            protocol_fee_payment,
+            protocol_fee_shares,
+        } = self.apply_fee(vault_protocol, vault_equity, now)?;
+
+        let vault_shares_before: u128 = self.get_protocol_shares(vault_protocol);
+
+        let (withdraw_value, n_shares) = withdraw_unit.get_withdraw_value_and_shares(
+            withdraw_amount,
+            vault_equity,
+            self.get_protocol_shares(vault_protocol),
+            self.total_shares,
+        )?;
+
+        validate!(
+            n_shares > 0,
+            ErrorCode::InvalidVaultWithdrawSize,
+            "Requested n_shares = 0"
+        )?;
+
+        let total_vault_shares_before = self.total_shares;
+        let user_vault_shares_before = self.user_shares;
+
+        if let Some(vp) = vault_protocol {
+            vp.last_protocol_withdraw_request.set(
+                vault_shares_before,
+                n_shares,
+                withdraw_value,
+                vault_equity,
+                now,
+            )?;
+        }
+        self.total_withdraw_requested = self.total_withdraw_requested.safe_add(withdraw_value)?;
```

Comment:
does it make sense to throw earlier in this function if there's no `vault_protocol`? otherwise this may get erroneous inflated, or move this variable update into the if statement above